core/os/device/deviceinfo/cc/android/query.cpp

```diff
@@ -271,48 +265,39 @@ bool createContext(void* platform_data) {
 
 #undef CHECK
 
-#define CHECK(x)                                                              \
-  if (!x) {                                                                   \
-    snprintf(gContext.mError, sizeof(gContext.mError), "JNI error:\n   " #x); \
-    destroyContext();                                                         \
-    return false;                                                             \
-  }
-
-  JavaVM* vm = reinterpret_cast<JavaVM*>(platform_data);
-  JNIEnv* env = nullptr;
-  auto res = vm->GetEnv(reinterpret_cast<void**>(&env), JNI_VERSION_1_6);
-  bool shouldDetach = false;
-  switch (res) {
-    case JNI_OK:
-      break;
-    case JNI_EDETACHED:
-      res = vm->AttachCurrentThread(&env, nullptr);
-      if (res != 0) {
-        snprintf(gContext.mError, sizeof(gContext.mError),
-                 "Failed to attach thread to JavaVM. (%d)", res);
-        destroyContext();
-        return false;
-      }
-      shouldDetach = true;
-      break;
-    default:
-      snprintf(gContext.mError, sizeof(gContext.mError),
-               "Failed to get Java env. (%d)", res);
-      destroyContext();
-      return false;
-  }
+#define GET_PROP(name, trans)                            \
+  do {                                                   \
+    char _v[PROP_VALUE_MAX] = {0};                       \
+    if (__system_property_get(name, _v) == 0) {          \
+      snprintf(gContext.mError, sizeof(gContext.mError), \
+               "Failed reading property %s", name);      \
+      destroyContext();                                  \
+      return false;                                      \
+    }                                                    \
+    trans;                                               \
+  } while (0)
+
+#define GET_STRING_PROP(n, t) GET_PROP(n, t = _v)
+#define GET_INT_PROP(n, t) GET_PROP(n, t = atoi(_v))
+#define GET_STRING_LIST_PROP(n, t)      \
+  do {                                  \
+    std::string _l, _t;                 \
+    GET_STRING_PROP(n, _l);             \
+    std::istringstream _s(_l);          \
+    while (std::getline(_s, _t, ',')) { \
+      t.push_back(_t);                  \
+    }                                   \
+  } while (0)
 
   std::string manufacturer;
   std::string model;
 
-  Class build(env, "android/os/Build");
-  CHECK(build.get_field("SUPPORTED_ABIS", gContext.mSupportedABIs));
-  CHECK(build.get_field("HOST", gContext.mHost));
-  CHECK(build.get_field("SERIAL", gContext.mSerial));
-  CHECK(build.get_field("MANUFACTURER", manufacturer));
-  CHECK(build.get_field("MODEL", model));
-  CHECK(build.get_field("HARDWARE", gContext.mHardware));
-  CHECK(build.get_field("DISPLAY", gContext.mOSBuild));
+  GET_STRING_LIST_PROP("ro.product.cpu.abilist", gContext.mSupportedABIs);
```

Comment:
Be sure to test this on older versions of Android and with 3rd party devices. IIRC, there was something preventing me from using system properties to get this info, hence the JNI.

Reply:
According to https://developer.android.com/about/versions/nougat/android-7.0-changes `__system_property_get` is the function to call. The warning there is simply that there is no standard specifying the names and values of the existing properties. However, the properties I'm querying here are the same as the ones queries by `android.os.Build`. BTW, the native code behind `android.os.Build` also uses the `__system_property_*` functions.

Reply:
I don't think the issue was with the property names, more that the function was simply not available on all versions of Android.
To be honest, I'm slightly surprised that this is accessible at all, as I thought there was an effort to classify this API as private (non-NDK-public). These release notes appear to suggest otherwise though.

Reply:
I think you are probably referring to this: https://stackoverflow.com/questions/28413530/api-to-get-android-system-properties-is-removed-in-arm64-platforms
As far as I can make out, it may at one point have been the idea to remove them, but that idea seems to have been overturned. There is a lot of Android code that depends on those functions to be present and accessible to apps. With those release notes, they are now guaranteed at the API level we build for.

Reply:
from the 7.0 docs, I would agree with Pascal that these are now officially supported only because of the wording:

```
Use __system_property_get instead of the private property_get
```
Which implies that __system_property_get is public. So I think this should be fine.